oauth_tools/oauth_helpers.py

```diff
@@ -0,0 +1,218 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import re
+from os.path import join
+from typing import Dict, List, Optional
+
+from playwright.async_api import expect
+from playwright.async_api._generated import BrowserContext, Page
+from pytest_operator.plugin import OpsTest
+
+from oauth_tools.constants import APPS
+from oauth_tools.external_idp import DexIdpManager, ExternalIdpManager
+
+logger = logging.getLogger(__name__)
+
+
+async def get_reverse_proxy_app_url(
+    ops_test: OpsTest, ingress_app_name: str, app_name: str
+) -> str:
+    """Get the address of a proxied application."""
+    status = await ops_test.model.get_status()  # noqa: F821
+    address = status["applications"][ingress_app_name]["public-address"]
+    return f"https://{address}/{ops_test.model.name}-{app_name}/"
+
+
+async def deploy_identity_bundle(
+    ops_test: OpsTest,
+    bundle_url: str = "identity-platform",
+    bundle_channel: Optional[str] = None,
+    ext_idp_service: Optional[ExternalIdpManager] = None,
+):
+    """Helper function for deploying and configuring the identity bundle and its dependencies."""
+    if ext_idp_service and not isinstance(ext_idp_service, ExternalIdpManager):
+        raise ValueError(
+            f"Invalid ext_idp_service type: {type(ext_idp_service)}, MUST be ExternalIdpManager or None"
+        )
+
+    deploy_cmd = ["juju", "deploy", bundle_url, "--trust"]
+    if bundle_channel:
+        deploy_cmd.extend(["--channel", bundle_channel])
+    await ops_test.run(*deploy_cmd)
+
+    # Wait for apps to go active, kratos_external_idp_integrator needs config to unblock
+    await ops_test.model.wait_for_idle(
+        [getattr(APPS, k) for k in APPS._fields if k != "KRATOS_EXTERNAL_IDP_INTEGRATOR"],
+        raise_on_blocked=False,
+        status="active",
+        timeout=2000,
+    )
+
+    if not ext_idp_service:
+        return
+
+    await ops_test.model.applications[APPS.KRATOS_EXTERNAL_IDP_INTEGRATOR].set_config(
+        {
+            "client_id": ext_idp_service.client_id,
+            "client_secret": ext_idp_service.client_secret,
+            "provider": "generic",
+            "issuer_url": ext_idp_service.issuer_url,
+            "scope": "profile email",
+            "provider_id": "Dex",
+        }
+    )
+    await ops_test.model.wait_for_idle(
+        raise_on_blocked=False,
+        status="active",
+        timeout=2000,
+    )
+
+    get_redirect_uri_action = (
+        await ops_test.model.applications[APPS.KRATOS_EXTERNAL_IDP_INTEGRATOR]
+        .units[0]
+        .run_action("get-redirect-uri")
+    )
+
+    action_output = await get_redirect_uri_action.wait()
+    assert "redirect-uri" in action_output.results
+
+    ext_idp_service.update_redirect_uri(redirect_uri=action_output.results["redirect-uri"])
+
+
+async def clean_up_identity_bundle(
```

Comment:
Just a **NONBLOCKING** question: (just a random thought) do you think it is worth we provide an additional async context manager for other teams to use so that they don't need to remember to call something to do resource cleanup after deploying the bundle?